vars/promotionHelper.groovy

```diff
@@ -29,22 +33,29 @@ void abortPreviousBuilds(RunWrapper promotedBuild) {
  *
  * @param promotedBuild the downstream build instance to be queried
  * @param upstreamJobName the upstream build name
+ * @param instance Hudson instance as returned by `Jenkins.instance`. Mostly useful to override when testing.
  * @return the build number from an upstream job
  */
-static int getUpstreamTriggerOrLatestBuildNumber(RunWrapper promotedBuild, String upstreamJobName) {
+static int getUpstreamTriggerOrLatestBuildNumber(RunWrapper promotedBuild, String upstreamJobName, Hudson instance = null) {
     int buildNum
+    def build
 
     // if build is trigger from upstream, retrieve the build number of the build that triggered this build
-    promotedBuild.getUpstreamBuilds().each { b ->
-        buildNum = b.number
-        println(">> Automatic trigger detected, using build ${b.getFullDisplayName()}")
+    def upstreamBuilds = promotedBuild.getUpstreamBuilds()
+    if (!upstreamBuilds.isEmpty()) {
+        build = upstreamBuilds.first()
+        println(">> Automatic trigger detected, using build ${build.getFullDisplayName()}")
+        buildNum = build.number
         return buildNum
     }
 
     // if build is trigger manually by a user, there is no upstream, so get the latest build number
-    def job = Jenkins.instance.getJob(upstreamJobName)
-    def build = job?.lastSuccessfulBuild
-    if (build == null) {
+    if (null == instance) {
```

Comment:
not java-y, should be "instance == null"

Reply:
![image](https://user-images.githubusercontent.com/19319040/80564630-aab15380-8a31-11ea-98b3-490761d890c2.png)

There's way too many arguments about whether there is any benefit in Java any more, so I'll just put it back on the right ;)